### PR TITLE
Allow renaming/creating/deleting files in a workspaced edit

### LINKED ISSUE #10659: WorkspaceEdit should support creating, deleting, and renaming of files
- VSCode Version: 1.4

Languages like Java require that renaming the top level class needs to rename the file as well. Currently the rename provider doesn't allow me to do that.

### PR DESCRIPTION
**Pull request by mjbvz**
Fixes #10659

Allows workspace edits to also change files in the workspace

### CODE REVIEW

File: src/vs/editor/common/modes.ts [hunk 1]
```diff
@@ -821,8 +821,22 @@ export interface IResourceEdit {
...
 }
+
+export interface IResourceRename {
+	readonly from: URI;
+	readonly to: URI;
+}
+
+export interface IResourceCreate {
+	readonly uri: URI;
+	readonly contents: string;
```
Comment (gorkem) [inline src/vs/editor/common/modes.ts]: Is `contents` field really necessary? WorkspaceEdit also has edits that can be applied to 0,0 location.

File: src/vs/editor/browser/services/bulkEdit.ts [hunk 2]
```diff
@@ -18,6 +18,13 @@ import { Selection, ISelection } from 'vs/editor/common/core/selection';
 import { IIdentifiedSingleEditOperation, ITextModel, EndOfLineSequence } from 'vs/editor/common/model';
 import { IProgressRunner } from 'vs/platform/progress/common/progress';
 import { ICodeEditor } from 'vs/editor/browser/editorBrowser';
+import { IResourceRename, IResourceCreate } from 'vs/editor/common/modes';
+
+export interface IResourceFileEdit {
+	readonly renamedResources: { from: URI, to }[];
```
Comment (jrieken) [inline src/vs/editor/browser/services/bulkEdit.ts]: `:URI`

File: src/vs/editor/common/modes.ts (hunk 1 above)
Comment (jrieken) [reply src/vs/editor/common/modes.ts]: Yeah, tempting... We could then just have one type, a rename is `from: Uri, to: Uri`, a create would be `from: null, to: Uri`, and a delete would be `from: Uri, to: null`... Similar to how we map inserted, deleted, and changed text

**Comment by jrieken** [comment]
@gorkem Just wondering in what order JDT is generating edit when renaming a class? Is it first a file rename and then an edit in the new file or the other way around, text edit the old and then rename? Or would expect to have both supported (which our proposal doesn't)

**Comment by gorkem** [comment]
@jrieken At the framework level both orders are supported. Everything is seen as a `Change` whether it is a `TextChange` or a `ResourceChange` so it applies them on the given order and trusts that the entity that created the chages knows the best order for it. This is a bit complex to do on WorkspaceEdit at the moment as TextEdits and Resource changes are treated separately.

**Comment by jrieken** [comment]
Makes sense @gorkem. I will give it a shot and see what I can do there... Today (without resources changes) we group all changes by file and performs per file... I'll try to preserve that but with interleaving resource changes...

**Comment by jrieken** [comment]
@mjbvz I have pushed quite a monster commit to support more interleaving between text and file changes. I say 'more' because some are extra tricky because of how the API is designed, the workspace edit automatically groups text edits by resource. That mean the followin

1. text change in `a`
2. rename `a` to `b`
3. another text change in `a`

will be executed in this order: `1, 3, 2`. I think that's a fair compromise tho (and I'll document that properly). I have removed most check and I rely on the subcomponents fails, e.g. when making a textual change to a deleted/non-existing file you'll get an error.

File: src/vs/workbench/api/node/extHostTypes.ts [hunk 3]
```diff
@@ -491,10 +491,29 @@ export class TextEdit {
...
 }
 
-export class WorkspaceEdit {
+export class WorkspaceEdit implements vscode.WorkspaceEdit {
 
-	private _values: [URI, TextEdit[]][] = [];
-	private _index = new Map<string, number>();
+	private _clock: number = 0;
+
+	private _resourceEdits: [number/*time*/, URI, URI][] = [];
```
Comment (mjbvz) [inline src/vs/workbench/api/node/extHostTypes.ts]: Same comment about avoid tuples in our internals as well. They are weird type system wise (`typeof([1, 2]) !== type [number, number]`) and it easy to confuse tuple fields when writing or refactoring code

**Review by mjbvz — COMMENTED** [review]
I scanned over the changes but am not familiar enough with the editor logic to give change around `bulkEdit` a proper review.

The compromise around edit ordering you suggest sounds reasonable, although I'm not so sure if is something we should support. How much complexity does supporting out of order edits add to the implementation? Could the first iteration of this API have extra checks that prevent constructing an out of order workspaceEdit and then we could relax this in the future if the restriction is causing problems for ext developers?

File: src/vs/vscode.d.ts [hunk 4]
```diff
@@ -2485,6 +2485,42 @@ declare module 'vscode' {
...
+		deleteResource(uri: Uri): void;
+
+		/**
+		 * Get the resource edits for this workspace edit.
+		 *
+		 * @returns A array of uri-tuples in which a rename-edit
+		 * is represented as `[from, to]`, a delete-operation as `[from, null]`,
+		 * and a create-operation as `[null, to]`;
+		 */
+		resourceEdits(): [Uri, Uri][];
```
[the remark this answers is not in the document]
Comment (BTaskaya) [reply src/vs/vscode.d.ts]: @mjbvz agree.

File: src/vs/workbench/api/node/extHostTypes.ts (hunk 3 above)
Comment (BTaskaya) [reply src/vs/workbench/api/node/extHostTypes.ts]: @mjbvz agree.

Comment (jrieken) [reply src/vs/workbench/api/node/extHostTypes.ts]: Yeah, can make that change internally.. For the API I'd go for consistency even tho I share the concern. Having said that, I don't expect heavy usage of those reading APIs

**Pull request merged**